document/js-api/index.bs

```diff
@@ -1115,6 +1148,122 @@ The algorithm <dfn>ToWebAssemblyValue</dfn>(|v|, |type|) coerces a JavaScript va
+            1. Otherwise,
+               1. Let |values| be « ».
+               1. [=list/iterate|For each=] |r| of |result|,
+               1. [=list/Append=] [=ToJSValue=](|r|) to |values|.
+               1. let |jsReturnValue| be [$CreateArrayFromList$](|values|).
+            1. Perform [=EvaluateCall=] (|promise|.\[[Resolve]],|jsReturnValue|,false)
+            1. Return UNDETERMINED
+    1. Let |con| be [=CreateBuiltinFunction=](|runner|,0,"",<<>>)
+    1. Perform [$Call$](|con|, undefined, <<>>)
+    1. Returns |promise|
+</div>
+
+Note: The extra |$Call$| in the above algorithm ensures that the creation of the Promise is separated from the fullfilling of that Promise. In effect, this allows suspension of the fullfillment to occur whilst allowing the creation of the Promise itself to continue.
+
+<div algorithm>
+    The <dfn constructor for="Suspending">Suspending(|jsFun|)</dfn> constructor, when invoked, performs the following steps:
+    1. If [$IsCallable$](|jsFun|) is false, throw a {{TypeError}}.
+    1. Let |suspendingProto| be \[[%WebAssembly.Suspending.prototype%]]
+    1. Let |susp| be the result of [$OrdinaryObjectCreate$](|suspendingProto|)
+    1. Assign the \[[wrappedFunction]] internal slot of |susp| to |jsFun|
+    1. Return |susp|
+</div>
+
+<div algorithm>
+To <dfn>create a suspending function</dfn> from a JavaScript function |func|, with type |functype| perform the following steps:
+
+    1. Assert: [$IsCallable$](|func|).
+    1. Let |stored settings| be the <a spec=HTML>incumbent settings object</a>.
+    1. Let |hostfunc| be a [=host function=] which performs the following steps when called with arguments |arguments|:
+        1. Let |realm| be |func|'s [=associated Realm=].
+        1. Let |relevant settings| be |realm|'s [=realm/settings object=].
+        1. Let |async_context| be the [=surrounding agent=]'s [=running execution context=].
+        1. [=Prepare to run script=] with |relevant settings|.
+        1. [=Prepare to run a callback=] with |stored settings|.
+        1. Let [|parameters|] → [|results|] be |functype|.
+        1. Let |jsArguments| be the result of [=coerce WebAssembly arguments=](|parameters|,|arguments|)
+        1. Let |ret| be [=?=] [$Call$](|func|, undefined, |jsArguments|).
+        1. [=Clean up after running a callback=] with |stored settings|.
+        1. [=Clean up after running script=] with |relevant settings|.
+        1. Assert: |ret|.\[[Type]] is <emu-const>throw</emu-const> or <emu-const>normal</emu-const>.
+        1. If |ret|.\[[Type]] is <emu-const>throw</emu-const>, then trigger a WebAssembly trap, and propagate |ret|.\[[Value]] to the enclosing JavaScript.
+        1. If [=list/size=] of |ret| is 1 and [$IsPromise$](|ret|.\[[Value]][0]):
+            1. Let |promise| be |ret|.\[[Value]][0].
+            1. Suspend |async_context|, removing it from the ambient [=Execution Context Stack=].
```

Comment:
I am deeply uncomfortable with this. I understand where @fgmccabe is coming from, and that the powers of the Suspend verb is underspecified in 262 itself. But I am not comfortable with the heavy lifting it's doing that it's able to effectively pause _any embedder spec of ecma262_. The current usage of Suspend in ecma262 itself is only thinks about suspension of 262 steps. If it had general suspension powers, then it would also be able to suspend in the middle of e.g. web API functions. That is not an implementable thing today.